### ISSUE TITLE
Get-DbaCmObject - PowerShellRemoting with SSL Not Working

### ISSUE THREAD

**Issue by fabiopennisifmg** [sqlcollaborative/dbatools#7613]
### Verified issue does not already exist?

No

### What error did you receive?

No error, unexpected behaviour.

VERBOSE: [14:50:46][Get-DbaCmObject] Configuration loaded | Cache disabled: False
VERBOSE: [14:50:46][Get-DbaCmObject] [xxx.xxx.local] Retrieving Management Information
WARNING: [14:50:46][Get-DbaCmObject] [xxx.xxx.local] Unable to find a connection to the target system. Ensure the name is typed correctly, and the server allows any of the following protocols: PowerShellRemoting | Multiple protocol connections were attempted, but no successful connections could be established with the specified computer.

### Steps to Reproduce

```powershell

#test connection using "native" commands - to show it works outside of dbatools
Test-WSMan -ComputerName $ComputerName -UseSSL -Port 5986 -Verbose
Invoke-Command -ComputerName $ComputerName -UseSSL -Port 5986 -ScriptBlock {Get-WMIObject Win32_ComputerSystem}

#config dbatools to usessl  and only allow powershell remoting protocol
Set-DbatoolsConfig -FullName 'psremoting.sessions.enable' -Value $true
Set-DbatoolsConfig -FullName 'psremoting.pssession.usessl' -Value $true -Register
Set-DbatoolsConfig -FullName 'computermanagement.type.disable.cimrm' -Value $true -Register
Set-DbatoolsConfig -FullName 'computermanagement.type.disable.cimdcom' -Value $true -Register
Set-DbatoolsConfig -FullName 'computermanagement.type.disable.wmi' -Value $true -Register
Set-DbatoolsConfig -FullName 'computermanagement.type.disable.powershellremoting' -Value $false -Register

#test dbatools commands
Get-DbaCmObject -ClassName Win32_ComputerSystem -ComputerName $ComputerName -Verbose
Get-DbaComputerSystem -ComputerName $ComputerName -Verbose
```

### Are you running the latest release?

v1.1.0

### Other details or mentions

We are trying to perform WinRM queries using HTTPS (port 5986) and not the default HTTP (5985).

Recently (see bug #7473 and task #7477) the "PowerShellRemoting" code region in Get-DbaCmObject was kindly adapted to use Invoke-Command2, which considers the 'psremoting.pssession.usessl' config value. 

We've looked at the code a few times in Get-DbaCmObject, by configuring 'psremoting.pssession.usessl' to true and disabling all 'computermanagement.type.disable.*' except 'computermanagement.type.disable.powershellremoting' which is enabled, we should be able to force calls to Get-DbaCmObject via the "PowerShellRemoting" the code region. 

This then uses Invoke-Command2 and should honour our 'psremoting.pssession.usessl' config value, thus forcing traffic to use SSL (HTTPS - port 5986) - yet it does not seem work.

All we get back are warnings and blanks in the returned data.

Is anyone able to see any holes in my config or possibly the code that prevents this from working?

Thanks so much for your help.

### What PowerShell host was used when producing this error

Windows PowerShell (powershell.exe), Windows PowerShell ISE (powershell_ise.exe)

### PowerShell Host Version

Name                           Value                                                                                                         
----                           -----                                                                                                         
PSVersion                      5.1.19041.1023                                                                                                
PSEdition                      Desktop                                                                                                       
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}                                                                                       
BuildVersion                   10.0.19041.1023                                                                                               
CLRVersion                     4.0.30319.42000                                                                                               
WSManStackVersion              3.0                                                                                                           
PSRemotingProtocolVersion      2.3                                                                                                           
SerializationVersion           1.1.0.1  

### SQL Server Edition and Build number

N/A

**Comment by potatoqualitee** [comment]
hey @fabiopennisifmg - can you confirm that you're using dbatools 1.1.0?

**Comment by fabiopennisifmg** [comment]
Hey @potatoqualitee, thanks for reaching out. Yup got version 1.1.0

**Comment by potatoqualitee** [comment]
Thank you! I'll ask @andreasjordan to take a look.

**Comment by andreasjordan** [comment]
I was able to recreate the problem. The fix works for me.
As it is only one line you can try to do the change in your local allcommands.ps1 in line 21331 and retest.
Please always open a new PowerShell, as information about failed connections are cached.

**Comment by andreasjordan** [comment]
@FriedrichWeinmann - While working on this, I found this as the initial setting:
```
# Disables or enables globally which Remote Management channels can be used
Set-DbatoolsConfig -FullName 'ComputerManagement.Type.Disable.CimRM' -Value $false -Initialize -Validation bool -Handler { [Sqlcollaborative.Dbatools.Connection.ConnectionHost]::DisableConnectionCimRM = $args[0] } -Description 'Globally disables all connections using Cim over WinRM'
Set-DbatoolsConfig -FullName 'ComputerManagement.Type.Disable.CimDCOM' -Value $false -Initialize -Validation bool -Handler { [Sqlcollaborative.Dbatools.Connection.ConnectionHost]::DisableConnectionCimDCOM = $args[0] } -Description 'Globally disables all connections using Cim over DCOM'
Set-DbatoolsConfig -FullName 'ComputerManagement.Type.Disable.WMI' -Value $true -Initialize -Validation bool -Handler { [Sqlcollaborative.Dbatools.Connection.ConnectionHost]::DisableConnectionWMI = $args[0] } -Description 'Globally disables all connections using WMI'
Set-DbatoolsConfig -FullName 'ComputerManagement.Type.Disable.PowerShellRemoting' -Value $true -Initialize -Validation bool -Handler { [Sqlcollaborative.Dbatools.Connection.ConnectionHost]::DisableConnectionPowerShellRemoting = $args[0] } -Description 'Globally disables all connections using PowerShell Remoting'
```

So only `CimRM` and `CimDCOM` are enabled, `WMI` and `PowerShellRemoting` are disabled. Can you tell me more about the reason they are disabled?

**Comment by FriedrichWeinmann** [comment]
WMI is plain old Get-WmiObject
Uses the same Tech as CimDCOM, but worse data and different client libraries.
You'd only want to use that if you have .NET Framework trouble. Also won't work on PS6+

PowerShellRemoting is using Invoke-Command to execute Get-WmiObject on the remote host.
This would only be viable if you must use WinRM but the CIM Component on the remote system is unable to talk with the target server's WinRM service.
From a protocol perspective, it uses the same as CimRM, but has a significantly higher resource impact on the target system.

So, both options are essentially fallback options in case of admittedly unlikely technical difficulties.

**Comment by fabiopennisifmg** [comment]
Hey @andreasjordan, 
Yes that fixes it for us. Thanks for looking at it.

To clarify what is happening here...
At present, there is no way to force a call with Get-DbaCmObject to use CimRM with SSL (HTTPS - WinRM port 5986). In my original bug/request (#7473) I incorrectly assumed that the 'psremoting.pssession.usessl' set to true would force the use of SSL and therefore WinRM HTTPS. @wsmelton correctly pointed out this was not the intention of the config value.

His suggest was to adapt the 'PowersShellRemoting' code region in the Get-DbaCmObject to use Invoke-Command2 (instead of Invoke-Command) as this would consider the 'psremoting.pssession.usessl' flag. This change was made in #7477. And it makes sense in my mind.

So... With that change in mind, to get the Get-DbaCmObject to communicate with HTTPS, the only path is using 'PowerShellRemoting'. To force that, we need to disable the computer management types CimRM, CimDCOM & WMI as none of these use or allow port 5986 (WinRM HTTPS port). However by enabling only PowerShellRemoting (together with the fix  @andreasjordan has described above) we can now make WinRM HTTPS queries.

Ideally we would like to force traffic via CimRM (which is the default) but then it needs to cater for SSL (HTTPS). It's only going through on 5985 - WinRM HTTP.

We are happy to use PowerShellRemoting for now, however, let me know if any of you reckon it's worth raising a request to adapt the CimRM code region in Get-DbaCmObject to allow HTTPS too.

Thanks again guys (@potatoqualitee, @FriedrichWeinmann, @andreasjordan & @wsmelton)!

**Comment by potatoqualitee** [comment]
Thanks for the confirmation! The fix is now included in 1.1.1 💯

**Comment by potatoqualitee** [comment]
@ClaudioESSilva -- this will be important to mention in the book.
